GeneticAlgorithm.java

```diff
@@ -0,0 +1,186 @@
+    private static int islandWeightIndex = 4;
+    private static int parityWeightIndex = 5;
+
+
+    public ArrayList<Genome> initializePopulation () {
+        population = new ArrayList<>();
+        //initial random weights for each feature
+        for (int i = 0; i < populationSize; i++) {
+            int id = i;
+            Genome genome = new Genome(id);
+            population.add(genome);
+        }
+        return population;
+    }
+
+    /*
+     * evaluate sets of weights
+     * first we need to evalute weights for every individual of the population
+     */
+    private static double evaluateIndividual (Genome individual, State s) {
+        //put weights into the array for evaluation 
+        double[] weights = new double[6];
+        weights[blockadeWeightIndex] = individual.getBlockadeWeight();
+        weights[edgeHeightWeightIndex] = individual.getEdgeHeightWeight();
+        weights[heightDifferenceWeightIndex] = individual.getHeightDifferenceWeight();
+        weights[holesWeightIndex] = individual.getHolesWeight();
+        weights[islandWeightIndex] = individual.getIslandWeight();
+        weights[parityWeightIndex] = individual.getParityWeight();
+
+        double score = 0;
+        new TFrame(s);
+        PlayerSkeleton p = new PlayerSkeleton();
+        while(!s.hasLost()) {
+            s.makeMove(p.pickMove(s,s.legalMoves(), weights));
+            s.draw();
+            s.drawNext(0,0);
+            try {
+                Thread.sleep(300);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }
+        score = s.getRowsCleared();
+
+        return score;
+    }
+
+    /*
+     * give score to each set of weights
+     * return the arraylist of pair(id, score): id: id for that set of weights; score: evaluation score
+     */
+    public static HashMap<Integer, Double> evaluatePopulation (ArrayList<Genome> population, State s) {
+        HashMap<Integer, Double> evaluationScore = new HashMap<Integer, Double>();
+
+        //evaluate each individual in the population.
+        //fixed index from a set of weights to its evaluation score
+        for (int i = 0; i < population.size(); i++) {
+            /* Map<id, score> */
+            State newState = new State();
+            evaluationScore.put(i, evaluateIndividual(population.get(i), newState));
+        }
+
+        return evaluationScore;
+    }
+
+    /*
+     * update population, each time choose the first and second sets of weight 
+     * their child will replace one of the parents
+     * return an updated population
+     */
+    public static ArrayList<Genome> evolve (ArrayList<Genome> population, HashMap<Integer, Double> evaluationScore) {
+        /* selection of parents */
+        int[] parents = selectParent(evaluationScore);
+        int mom = parents[0];
+        int dad = parents[1];
+
+        Genome momGenome = population.get(mom);
+        Genome dadGenome = population.get(dad);
```

Comment:
You can consider creating several children from a mixed bag of parents instead of just one child per generation. to speed up the learning process. (e.g. keep the top half of the population, make them mate and replace the bottom half of the population with children)

Reply:
Yes, i realise this can be a potential problem. I am considering selecting 10% of the population to produce children and replace half of the population.